### PR TITLE
Kokoro Build Script to Run Pytorch Model on NVIDIA Container

### PR DESCRIPTION
**Pull request by raj-prince**
>> Have to resolve issues:
(a) Log-deletion haven't added in the script.
(b) Overall testing is in progress.

### CODE REVIEW

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+nohup /pytorch_dino/gcsfuse/gcsfuse --type-cache-ttl=1728000s \
+        --stat-cache-ttl=1728000s \
+        --stat-cache-capacity=1320000 \
+        --stackdriver-export-interval=60s \
+        --implicit-dirs \
+        --disable-http2 \
+        --max-conns-per-host=100 \
+        --debug_fs \
+        --debug_gcs \
+        --log-file logs.txt \
```
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh:23]: Does it include log rotation implicitly (i.e. no need to install cron job?) ? If yes, we can add comments telling about size of each log file, duration of logs and how many hours of logs we keep.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+folder_file="/opt/conda/lib/python3.8/site-packages/torchvision/datasets/folder.py"
+x=$(grep -n "def pil_loader(path: str) -> Image.Image:" $folder_file | cut -f1 -d ':')
+y=$(grep -n "def accimage_loader(path: str) -> Any:" $folder_file | cut -f1 -d ':')
+y=$((y - 2))
+lines="$x,$y"
+sed -i "$lines"'d' $folder_file
+sed -i "$x"'r bypassed_code.py' $folder_file
+
+# Fix the caching issue, by downloading the issue
```
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh:47]: I didn't get it from comment. Can we elaborate this comment ?

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh (hunk 1 above)
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh:23]: nit: create a directory with name 'logs' inside gcsfuse directory or directory with name 'gcsfuse' alongside 'gcsfuse_data' and store gcsfuse logs inside it for better organisation.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+
+# Steps to bootstrap with gpu docker
+# Install docker
```
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh:4]: nit: remove or change this comment to something like:
"Steps to install nvidia-driver, docker-engine, nvidia-docker2 and their required dependencies"

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+sudo apt-get install docker-ce docker-ce-cli containerd.io docker-compose-plugin -y
+
+
+echo "Installing driver..."
+sudo apt update && sudo apt install -y build-essential
+BASE_URL=https://us.download.nvidia.com/tesla
+DRIVER_VERSION=450.172.01
+sudo curl -fSsl -O $BASE_URL/$DRIVER_VERSION/NVIDIA-Linux-x86_64-$DRIVER_VERSION.run
+
+sudo sh NVIDIA-Linux-x86_64-$DRIVER_VERSION.run -s
```
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh:32]: Question: Does this command take care of choosing appropriate options in the prompt that comes up ? If yes, does the '-s' option at the end takes care of it ?

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+python -c 'import torch;torch.hub.list("facebookresearch/xcit:main")'
+
+# Run the pytorch Dino model
+# We need to run it in foreground mode to make the container running.
+# TODO: Please reset the value of gpu according to the availability
```
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh:52]: Leaving a comment here, so that TODO is taken care of.

**Review by sethiay — COMMENTED** [review]
Looks good with some minor comments.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
+
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
+  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+
+sudo apt-get update
+sudo apt-get install docker-ce docker-ce-cli containerd.io docker-compose-plugin -y
+
+
```
Comment (raj-prince) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh:25]: Also, please add the steps to install the ops-agent and also make sure to run gcsfuse with stack-driver agent.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh (hunk 3 above)
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh:4]: Can we also install ops agent ?

Comment (raj-prince) [reply perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh:4]: Yes, I have added a comment for that.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh [hunk 7]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+wget -O go_tar.tar.gz https://go.dev/dl/go1.19.4.linux-amd64.tar.gz
+rm -rf /usr/local/go && tar -C /usr/local -xzf go_tar.tar.gz
+export PATH=$PATH:/usr/local/go/bin
+
```
[the remark this answers is not in the document]
Comment (raj-prince) [reply perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh:6]: Done.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh (hunk 4 above)
Comment (raj-prince) [reply perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_host.sh:37]: Yes, we choose appropriate option by default. 
One of the option we chose - Install NVIDIA's 32-bit compatibility libraries? as yes - which seems fine. Other than that we only press ok for warning and installation completion.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh (hunk 1 above)
Comment (raj-prince) [reply perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh]: Added the cron job for this.

File: perfmetrics/scripts/continuous_test/ml_tests/build_dino.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+cd "${KOKORO_ARTIFACTS_DIR}/github/gcsfuse/perfmetrics/scripts"
+
+echo "Setting up the machine with Docker and Nvidia Driver"
+chmod +x ml_tests/setup_host.sh
+source ml_tests/setup_host.sh
+
+cd "${KOKORO_ARTIFACTS_DIR}/github/gcsfuse"
+echo "Building docker image containing all pytorch libraries..."
+sudo docker build . -f perfmetrics/scripts/ml_tests/pytorch_dino_model/Dockerfile --tag pytorch-gcsfuse
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/continuous_test/ml_tests/build_dino.sh:11]: is this docker file specific to dino_model or pytroch?

either ways, create a pytorch folder and then dino folder inside it.  same for build_dino.sh also.

File: perfmetrics/scripts/ml_tests/setup_host.sh [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+echo "Installing NVIDIA container tool..."
+distribution=$(. /etc/os-release;echo $ID$VERSION_ID) \
+      && curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
+      && curl -s -L https://nvidia.github.io/libnvidia-container/$distribution/libnvidia-container.list | \
+            sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
+            sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
+
+sudo apt-get update
+sudo apt-get install -y nvidia-docker2
+sudo systemctl restart docker
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/ml_tests/setup_host.sh:48]: please add a readme file specifying how is the pytorch dino model getting executed and what does each file represent.

**Review by vadlakondaswetha — CHANGES_REQUESTED** [review]

File: perfmetrics/scripts/continuous_test/ml_tests/build_dino.sh [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+mkdir container_artifacts
+
+echo "Running the docker image build in the previous step..."
+sudo docker run --runtime=nvidia --name=pytorch_automation_container --privileged -d -v ${KOKORO_ARTIFACTS_DIR}/github/gcsfuse/container_artifacts:/pytorch_dino/run_artifacts:rw,rshared \
+--shm-size=128g pytorch-gcsfuse:latest
+
+# Wait for the script completion as well as logs output.
+sudo docker logs -f pytorch_automation_container
```
Comment (sethiay) [inline perfmetrics/scripts/continuous_test/ml_tests/build_dino.sh:20]: Add code to copy logs of the run to bucket.

File: perfmetrics/scripts/continuous_test/ml_tests/continuous.cfg [hunk 11]
```diff
@@ -1,2 +1,2 @@
-build_file: "gcsfuse/perfmetrics/scripts/continuous_test/ml_tests/build.sh"
-timeout_mins: 5700
+build_file: "gcsfuse/perfmetrics/scripts/continuous_test/ml_tests/build_dino.sh"
+timeout_mins: 20160
```
Comment (sethiay) [inline perfmetrics/scripts/continuous_test/ml_tests/continuous.cfg:2]: we can add a comment that it runs for 14 days.

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+# Create a directory for gcsfuse logs
+mkdir gcsfuse_logs
+
+echo "Mounting GCSFuse..."
+nohup /pytorch_dino/gcsfuse/gcsfuse --type-cache-ttl=1728000s \
+        --stat-cache-ttl=1728000s \
+        --stat-cache-capacity=1320000 \
+        --stackdriver-export-interval=60s \
+        --implicit-dirs \
```
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh:24]: Shall we use --experimental-enable-storage-client-library flag here ?

File: perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+python3 -m torch.distributed.launch \
+  --nproc_per_node=2 dino/main_dino.py \
+  --arch vit_small \
+  --num_workers 20 \
+  --data_path gcsfuse_data/imagenet/ILSVRC/Data/CLS-LOC/train/ \
+  --output_dir "./run_artifacts/$experiment" \
+  --norm_last_layer False \
+  --use_fp16 False \
+  --clip_grad 0 \
+  --epochs 2 \
```
Comment (sethiay) [inline perfmetrics/scripts/ml_tests/pytorch_dino_model/setup_container.sh:97]: Adding comment to change it back to 800 epochs and nproc_per_node=8.

**Review by sethiay — CHANGES_REQUESTED** [review]

File: perfmetrics/scripts/continuous_test/ml_tests/build_dino.sh [hunk 14]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+cd "${KOKORO_ARTIFACTS_DIR}/github/gcsfuse/perfmetrics/scripts"
```
Comment (raj-prince) [inline perfmetrics/scripts/continuous_test/ml_tests/build_dino.sh:3]: Add error handling in all the bash file.

**Comment by raj-prince** [comment]
Created an another PR  https://github.com/GoogleCloudPlatform/gcsfuse/pull/900
This was required due to rebasing in the this PR. Closing it.

**Pull request closed without merging**